source/common/router/header_parser.cc

```diff
@@ -214,45 +211,59 @@ HeaderFormatterPtr parseInternal(const envoy::config::core::v3::HeaderValue& hea
     return std::move(formatters[0]);
   }
 
-  return std::make_unique<CompoundHeaderFormatter>(std::move(formatters), append);
+  return std::make_unique<CompoundHeaderFormatter>(std::move(formatters));
 }
 
 } // namespace
 
-HeaderParserPtr HeaderParser::configure(
-    const Protobuf::RepeatedPtrField<envoy::config::core::v3::HeaderValueOption>& headers_to_add) {
+HeaderParserPtr
+HeaderParser::configure(const Protobuf::RepeatedPtrField<HeaderAppendOption>& headers_to_add) {
   HeaderParserPtr header_parser(new HeaderParser());
 
   for (const auto& header_value_option : headers_to_add) {
-    const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
-    HeaderFormatterPtr header_formatter = parseInternal(header_value_option.header(), append);
+    HeaderAppendAction append_action;
+
+    if (header_value_option.has_append()) {
+      // 'append' is set and ensure the 'append_action' value is equal to the default value.
+      if (header_value_option.append_action() != HeaderAppendOption::APPEND_IF_EXISTS_OR_ADD) {
+        throw EnvoyException("Both append and append_action are set and it's not allowed");
+      }
+
+      append_action = header_value_option.append().value()
+                          ? HeaderAppendOption::APPEND_IF_EXISTS_OR_ADD
+                          : HeaderAppendOption::OVERWRITE_IF_EXISTS_OR_ADD;
+    } else {
+      append_action = header_value_option.append_action();
```

Comment:
if `append` is false, then I don't see how we can accept a default action here which is to essentially append (`APPEND_IF_EXISTS_OR_ADD`). Maybe we should only interpret `append_action` if `append` is true, in which case the semantics seem to make more sense.

Reply:
`append` is optional bool. We will use `append_action` only when the we don't set `append`.

If `append` is set to false, the `OVERWRITE_IF_EXISTS_OR_ADD` will be used. 

So there should be no any behaviors change.

Reply:
Got it, makes sense.